### ISSUE TITLE
Common benchmarking infrastructure for Torchbench, TIMM and huggingFace

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#93767]
To monitor our progress according to benchmarks agreed upon in https://docs.google.com/document/d/1g9imM66GaDpYZGFJH8WEtYEssvv0_EniKsky7dup6iM/edit# we need to work on a common benchmarking infrastructure that would allow us to get comparable benchmarking results for the different backend technologies. 
`torchbench.py` benchmarking harness in torchdynamo provides a lot of options needed for benchmarking, provides an easy harness to run torchbench models and allows one to easily swap the backends to compare perf, however, it is missing both feature and model coverage, and we should close this gap. 
I think it would make sense to use torchbench.py as a basis for the benchmarking harness, but if we want to do something else entirely, this issue is to discuss various approaches.
Below I propose a list of tasks with brief comments that I think we should complete in order to have good model and feature coverage. 

- [x] Setup instructions. E.g. currently torchbench needs to be installed with its dependencies, and it's not a straightforward process for source builds of torch/torchaudio/torchvision etc. Which exact commands should be run to get results we are comparing etc - PR started pytorch/torchdynamo#577
- [x] Enable training option in the harness 
- [ ] Enable correctness testing for training at least for the models without dropout, enable option to set dropout to 0 for correctness testing
- [ ] Agreed upon list of TIMM and Huggingface models. Some of the TIMM models present in Torchbench I couldn't find in nvidia plots (e.g. nfnet, efficientdet). cc @ezyang @soumith @msaroufim @wconstab @bdhirsh @kevinstephano, @Chillee 
- [x] A way to run Huggingface 
- [x] and Timm models that are not necessarily in TorchBench. TorchBench already installs HF and Timm repos as part of its setup process, but models and model configs that are not part of torchbench are not easy to access
- [x] A way to control batch size in the benchmarks (currently we pick up whatever torchbench says, but, per discussion in document above, we should be able to adjust this parameter) pytorch/torchdynamo#581
- [x] A correct way of enabling mixed precision via amp (see e.g. here https://github.com/kevinstephano/simple_dl_models/blob/e1246e7cb130873f21974cd9ab924853a173a872/execution/execution_loop.py#L97), currently `torchbench.py` just casts model to half. In general, Kevin's "simple_dl_models" repo is a useful reference for simplified training loops that we'll likely need
- [ ] bfloat16 benchmarking (currently only fp16 is enabled)
- [x] channels-last, with a way to mark models for which it makes sense (e.g. TIMM) and skip models for which it doesn't (HF)
- [ ] add optimizer to training loop
- [ ] provide a way to independently compute speedups for forward, forward + backward and forward + backward + optimizer
- [ ] record absolute times in addition to speed-ups so we can more readily compare different flavors for the same model such as channels last and low precision
- [ ] allow for easier extraction of produced fx graphs and profiles (unfortunately, profiling with kineto is not compatible with cuda graphs, but even without cuda graphs it's still valuable). 
- [ ] validate that results obtained from torchbench agree with results nvidia is currently providing.

Distributed benchmarking is left out of scope for this issue, but once we have consistent infra for single GPU we'll move to multi. 
cc @csarofeen, @xwang233, @wconstab
